rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,440 @@
+- Start Date: 2022-06-13
+- RFC PR: [#64](https://github.com/inveniosoftware/rfcs/pull/64)
+- Authors: Zacharias Zacharodimos, Pablo Panero
+
+# Custom fields
+
+## Summary
+
+InvenioRDM's record and community data models might not contain all the fields needed to fit some cases. However, they can be extended using _custom fields_. This RFC outlines the implementation details.
+
+## Motivation
+
+Different instances of InvenioRDM are setup to fit different use cases. In order to fulfil those, these instances will need to enrich the base data model with domain specific fields. Thus, the need for a mechanism to add/support these custom fields in all levels of the existing data models (jsonschema, marshmallow, Elasticsearch mappings) and make them available to the end user in the upload and community creation forms.
+
+### Use Cases
+
+**As an instance administrator...**
+
+- I want to add a custom fields to the record's data model, and/or the community's data model.
+- I want users to be able to input values for those custom fields via the upload and community creation forms.
+- I want to add custom fields for internal use only (i.e. not visible to the users, nor available in the forms).
+- I want to be able to hide some fields in some communities. While the fields might still be present in the data model, they won't be shown/available for input. Alternatively, I want to be able to configure custom fields per community.
+
+**The following use cases apply to all the above points**, and focus on the how fields are defined and used.
+
+- I want to be able to define the type: primitive (string, number, date), complex (list, dictionary), and vocabularies (both generic and specific).
+- I want to be able to define required fields.
+
+**When displaying the custom fields in the forms...**
+
+- I want to define groups for my new fields.
+- I want to define in which order they will display inside those groups.
+- I want to add fields to already existing groups (e.g. Basic information).
+- I want to support translations (i18n) of the fields' label and description.
+
+**When displaying the custom fields of records...**
+
+- I want to define where and how the field will be shown in the landing/community page.
+- I want to define how the values will be serialized into different formats (e.g. Dublin Core, Darwin Core, DataCite).
+- I want to be able to filter my search results based on my new field.
+- I want to support translations (i18n) for the values of the custom fields.
+
+### Fist iteration limitations
+
+The first iteration of custom fields will be available in InvenioRDM v10. It targets the implementation of two vertical workflows, allowing to add free text custom fields (e.g. keywords, descriptions) and also based on generic vocabularies (e.g. defining a new "experiment" vocabulary). This will enable the implementation of the mechanisms to support custom fields, and allow developers to extend support for new types of fields (e.g. dates).
```

Comment:
What do you mean with "vertical" workflows? Difference with an "horizontal" one?

I would also add number and bool in the first iteration, should not be more complex than string...

Reply:
What I meant was more or less like horizontal vs vertical development (i.e. full layer vs E2E use cases). I will rephrase it.

As for the bool and number... While I agree it should be easy, it was not in the first plan and I'd like to first have things merged and working before committing in paper to more. @zzacharo WDYT?

Reply:
Honestly, we can say primitive type or a more user-friendly work and give examples such as text, number etc... Adding a bool or number is not that difficult but I would probably vaguely say it here and then fully document the list in the documentation!

Reply:
Added bool and number